### DISCUSSION TITLE
Need some ideas for reusable React components and utilities

### DISCUSSION THREAD

**Discussion by gremo** [directus/directus#5434] [Feature Requests]
Hi folks! I'm really excited about this project ❤️ , I'm trying to build a React library with some components and helpers to work with Directus SDK and Directus API in general. To be used with React and especially Next.js.

Right now it's just a (working) WIP: https://github.com/gremo/react-directus

I've implemented an hook, a `<DirectusImage>` and `<DirectusAsset>` component. Is there something else I can work on? Do you have some idea?

Thank you 👍

**Comment by Jonesus** [comment]
Nice job! Some ideas what I've needed / manually implemented a couple of times for my Next.js + Directus projects:

1. **Type-safe API**
    - Maybe the `<DirectusProvider>` could accept a schema that's manually typed according to the shape of one's data in Directus, or when we get the official TypeScript endpoint from Directus itself the provider could accept an interface generated by it?
    - Fetching deeply nested relational data (translations etc.) is also a bit painful when manually writing the queries to include specific fields, not sure if there's any reasonable abstractions around it though.
2. **More optimized API fetching**
    - Quite often it becomes really beneficial to use something like `swr` or `react-query` in a project that fetches data more often than just statically generating, and in those cases I feel that the given libraries can quite often even be used instead of a global state manager, if the global state's only function is to fetch and provide data across the site
3. **User authentication handling**
    - In a Next.js app it would be really nice to have a hook for handling the authentication state for users, e.g. checking if a user has the required privileges for accessing a route.
4. **Out-of-the-box support for Next.js Image**
    - The lazy-loading provided by Next.js's `<Image>` is really nice for basically any site, and I feel could even warrant its own component. I have a (somewhat opinionated and use-case-specific) implementation here for inspiration: https://github.com/Prodeko/prodeko-community/blob/main/src/components/AnimatedImage.tsx
5. **Some way to map a collection / collections to Next.js pages**
    - So far I've implemented a somewhat obscure custom routing logic based on a catch-all page and convoluted data parsing logics, where the existence of query path is checked against some Directus collections that have a `slug` field, and on a match the appropriate data is returned. Would be really great to figure out a cleaner way to do this somehow that would support i18n, dynamic paths configurable via Directus without rebuilding etc. How I've done it previously:
    https://github.com/Prodeko/prodeko-community/blob/main/src/pages/%5B%5B...slug%5D%5D.tsx
    https://github.com/Prodeko/prodeko-community/blob/8890d111d15f5b93ea94b760629d81ac460095a4/src/api/index.ts#L142

Happy to discuss the ideas here or at your repo, Directus +  Next.js is a stack I've been playing around with for a while and would love to improve upon :)

**Comment by azrikahar** [comment]
Heya! Thank you for taking the time to submit this request! 

It has been over **90 days**, and this discussion has not received at least **15 votes** from the community. This means that we don't feel like there's enough community interest to warrant further R&D into this topic at this time. :ice_cube:

This request will now be closed to keep our discussions tidy. Please reach out if you have any questions!

_For more information, see our [Feature Request Process](https://docs.directus.io/contributing/feature-request-process.html)._
